### PR TITLE
Keep config

### PR DESCRIPTION
**Pull request by rodinux**
## Problem

- *to resolve the issue #49, the upgrade will remove personalization in the config.local.php*

## Solution

- *I propose to keep the file config.local.php on the upgrade script without reinstall it from config*

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by rodinux** [comment]
@alexAubin @ericgaspar , do you think we can do this ? 
Or is better add a file `config.local.user.php` dedicated to personalization on the folder `__DATA__DIR__/data` ??

**Comment by alexAubin** [comment]
There is no straightforward way to address the issue, this is a shortcoming of the current strategy in packaging regarding app config
1) One the one hand, you want the new version of the app config to be applied in case there are new keys to be added etc ...
2) On the other hand, you don't want to overwrite manually modified files because the amin may have tweaked stuff

The "real" way to address this would be to keep the original config file somewhere and use like a git merge thing somehow, but for now we don't have the logistic for this

**Comment by rodinux** [comment]
> There is no straightforward way to address the issue, this is a shortcoming of the current strategy in packaging regarding app config
> 
>     1. One the one hand, you want the new version of the app config to be applied in case there are new keys to be added etc ...
> 
>     2. On the other hand, you don't want to overwrite manually modified files because the amin may have tweaked stuff
> 
> 
> The "real" way to address this would be to keep the original config file somewhere and use like a git merge thing somehow, but for now we don't have the logistic for this

Ok, 
what about doing like this ?
I added a file config here `$data_dir/data/config.local.user.php`
Then a line at the bottom of the file `$intall_dir/config.local.php`

```
/**
 * Chemin vers le fichier pour des configurations personnelles
 * qui ne sera pas écraser lors des mises à jour.
 */
require '__DATA__DIR__/data/config.local.user.php';
```

Like this we should keep the `config.local.php` safe and users can customize on the file in their `$data_dir/data/config.local.user.php`... For example, they could configure API key and user, isn't it ? or Nextcloud sync or wopi_url to use collabora or only_office...

**Comment by rodinux** [comment]
Ok, I found how makes it working... 
!testme

**Comment by rodinux** [comment]
!testme

**Comment by rodinux** [comment]
Ok, So I have added some stuff when try 
* installing app `garradin_ynh`,
* made a pre-install, then a migration to paheko with the command `sudo yunohost app upgrade garradin -u https://github.com/YunoHost-Apps/paheko_ynh/tree/garradin-migration --debug`. 
* At this moment the migration is well, but not very clean and have old configurations with some files `$install_dir/config.local.yunohost.php` and `$install_dir/config.local.user.php` and also a duplicated directory `$install_dir/data/data`, so I do a first cleaning here:

https://github.com/YunoHost-Apps/paheko_ynh/blob/9c0aa9a58481d9a23316fd1d1aee68a6fa0e474a/scripts/upgrade#L35C1-L44C3

* Also, at the end of the upgrade to I need mv the db files `.sqlite` on the new path:

https://github.com/YunoHost-Apps/paheko_ynh/blob/9c0aa9a58481d9a23316fd1d1aee68a6fa0e474a/scripts/upgrade#L109-L112

This works nice...

**Comment by rodinux** [comment]
!testme

**Comment by rodinux** [comment]
!testme

**Comment by rodinux** [comment]
!testme

File: conf/config.local.php [hunk 1]
```diff
@@ -2,12 +2,13 @@
 
 /**
  * Ce fichier représente un exemple des constantes de configuration
- * disponibles pour Paheko.
+ * disponibles pour Paheko. Il est bâti depuis une copie du fichier '$install_dir/config.dist.php'
```
Comment (ericgaspar) [inline conf/config.local.php:5]:
```suggestion
 * disponibles pour Paheko. Il est bâti depuis une copie du fichier '__INSTALL_DIR__/config.dist.php'
```

Comment (rodinux) [reply conf/config.local.php:5]: Ok

**Comment by rodinux** [comment]
For me it's OK this time, after doing few tests...

File: scripts/upgrade [hunk 2]
```diff
@@ -29,6 +29,22 @@ if [ -z "${secret_key:-}" ]; then
 	ynh_app_setting_set --app=$app --key=secret_key --value=$secret_key
 fi
 
+#=================================================
+# KEEP DATA IF NEEDED
+#=================================================
+if ynh_compare_current_package_version --comparison eq --version 1.2.8~ynh1
```
Comment (rodinux) [inline scripts/upgrade:35]: if [ $(ynh_compare_current_package_version --comparison lt --version 1.3.0~ynh1) && $(ynh_compare_current_package_version --comparison ge --version 1.2.11~ynh5) ]
then
	ynh_script_progression --message="cleaning conf files..." --weight=5
	mv "$install_dir/config.local.php" "$install_dir/config.local.php.old"
	mv "$install_dir/config.local.yunohost.php" "$install_dir/config.local.php"
	chmod 650 "$install_dir/config.local.php"
	chown $app:$app "$install_dir/config.local.php"
	ynh_store_file_checksum --file="$install_dir/config.local.php"
	ynh_secure_remove "$install_dir/config.local.php.old"
	ynh_secure_remove "$install_dir/data/config.local.user.php"
	ynh_secure_remove "$install_dir/data/data"
fi

if ynh_compare_current_package_version --comparison lt --version 1.2.11~ynh5
then
	ynh_script_progression --message="cleaning conf files..." --weight=5
	mv "$install_dir/config.local.php" "$install_dir/config.local.php.old"
	mv "$install_dir/config.local.yunohost.php" "$install_dir/config.local.php"
	chmod 650 "$install_dir/config.local.php"
	chown $app:$app "$install_dir/config.local.php"
	ynh_store_file_checksum --file="$install_dir/config.local.php"
	ynh_secure_remove "$install_dir/config.local.php.old"
	ynh_secure_remove "$install_dir/config.local.user.php"
	ynh_secure_remove "$install_dir/data/data"
fi

**Comment by rodinux** [comment]
!testme

**Comment by rodinux** [comment]
> There is no straightforward way to address the issue, this is a shortcoming of the current strategy in packaging regarding app config
> 
>     1. One the one hand, you want the new version of the app config to be applied in case there are new keys to be added etc ...
> 
>     2. On the other hand, you don't want to overwrite manually modified files because the amin may have tweaked stuff
> 
> 
> The "real" way to address this would be to keep the original config file somewhere and use like a git merge thing somehow, but for now we don't have the logistic for this

Hello, after lot of tests, I think I can push this PR, it will keep a `config.local.user.php` for the users in the `$data_dir/data/` folder, needed for add some personal configurations and keep them on an upgrade. It keep the original config file also. 

Do you approve ? also @ericgaspar is it OK for you ?

**Comment by ericgaspar** [comment]
approved

**Comment by rodinux** [comment]
> but to be honest I am not really following the package :)

Ok thanks.

**Pull request merged**